migrations/20220830011729_group_configs__using_group_roles__add.sql

```diff
@@ -0,0 +1,10 @@
+ALTER TABLE group_configs ADD COLUMN using_group_roles boolean DEFAULT false;
+
+ALTER TABLE group_users DROP CONSTRAINT group_users_pkey;
+ALTER TABLE group_users ADD PRIMARY KEY (group_id, user_id, group_role_id);
+
+-- Insert role for group assessments that don't use group roles
+INSERT INTO group_roles 
+    (role_name, maximum)
+VALUES
+    ('No group roles', 99);
```

Comment:
Since our original solution required `group_users` have all three columns `group_id`, `user_id`, and `group_role_id` to be non-`NULL` as they were part of the primary key, we used this value to get around that.

I committed a change to create a new relational table `group_user_roles` and to revert `group_users` back to having only a `group_id` and `user_id`. See: https://github.com/PrairieLearn/PrairieLearn/pull/6109/commits/74baeda59120fef0d283ccc4653af27bb16907ba

We originally thought to implement roles this way, but we didn't want to duplicate information from `group_users` into another table. Is there a better way to go about implementing this?

Reply:
Probably the best way would have been to not use `(group_id, user_id)` as a primary key for `group_users`, but instead use an auto-incrementing ID. You could then have a `group_user_roles` table that references that ID (`group_user_id`).

That said, separating group membership and group roles into separate tables is already a much better way to go! And in fact, if you treat `(group_id, user_id)` as a `FOREIGN KEY` in the `group_user_roles` table, you won't *really* be duplicating any information, and you can take advantage of cascading deletes too. Let me know if you want me to expand on that.